internal/pkg/instrumentation/bpf/go.opentelemetry.io/auto/sdk/bpf/probe.bpf.c

```diff
@@ -0,0 +1,219 @@
+		bpf_printk("write_span_context: failed to write span ID: %ld", ret);
+	}
+
+	void *flags = (void *)(go_sc + span_context_trace_flags_pos);
+	ret = bpf_probe_write_user(flags, &sc->TraceFlags, TRACE_FLAGS_SIZE);
+	if (ret != 0) {
+		rc--;
+		bpf_printk("write_span_context: failed to write trace flags: %ld", ret);
+	}
+
+	return rc;
+}
+
+static __always_inline long span_context_from_Go(void *go_span_context, struct span_context *sc) {
+	if (go_span_context == NULL) {
+		bpf_printk("span_context_from_Go: empty span context");
+		struct span_context zero = {0};
+		*sc = zero;
+		return 0;
+	}
+
+	long rc = 0;
+	void *tid = (void *)(go_span_context + span_context_trace_id_pos);
+	long ret = bpf_probe_read(&sc->TraceID, TRACE_ID_SIZE, tid);
+	if (ret != 0) {
+		rc--;
+		bpf_printk("span_context_from_Go: failed to read trace ID", ret);
+	}
+
+	void *sid = (void *)(go_span_context + span_context_span_id_pos);
+	ret = bpf_probe_read(&sc->SpanID, SPAN_ID_SIZE, sid);
+	if (ret != 0) {
+		rc--;
+		bpf_printk("span_context_from_Go: failed to read span ID", ret);
+	}
+
+	void *flags = (void *)(go_span_context + span_context_trace_flags_pos);
+	ret = bpf_probe_read(&sc->TraceFlags, TRACE_FLAGS_SIZE, flags);
+	if (ret != 0) {
+		rc--;
+		bpf_printk("span_context_from_Go: failed to read trace flags", ret);
+	}
+
+	return rc;
+}
+
+SEC("uprobe/Tracer_start")
+int uprobe_Tracer_start(struct pt_regs *ctx) {
+    struct go_iface go_context = {0};
+    get_Go_context(ctx, 2, 0, true, &go_context);
+
+	struct otel_span_t otel_span;
+    __builtin_memset(&otel_span, 0, sizeof(struct otel_span_t));
+
+    start_span_params_t params = {
+        .ctx = ctx,
+        .go_context = &go_context,
+        .psc = &otel_span.psc,
+        .sc = &otel_span.sc,
+        .get_parent_span_context_fn = NULL,
+        .get_parent_span_context_arg = NULL, // Default to new root. 
+    };
+
+    start_span(&params);
+
+	void *parent_span_context = get_argument(ctx, 5);
+	long rc = write_span_context(parent_span_context, &otel_span.psc);
+	if (rc != 0) {
+		bpf_printk("failed to write parent span context: %ld", rc);
+	}
+
+	if (!is_sampled(params.sc)) {
+		// Default SDK behaviour is to sample everything. Only set the sampled
+		// value to false when needed.
+		void *sampled_ptr_val = get_argument(ctx, 6);
+		if (sampled_ptr_val == NULL) {
+			bpf_printk("nil sampled pointer");
+		} else {
+			bool false_val = false;
+			rc = bpf_probe_write_user(sampled_ptr_val, &false_val, sizeof(bool));
+			if (rc != 0) {
+				bpf_printk("bpf_probe_write_user: failed to write sampled value: %ld", rc);
+			} else {
+				bpf_printk("wrote sampled value");
+			}
+		}
+	}
+
+	void *span_context_ptr_val = get_argument(ctx, 7);
+	rc = write_span_context(span_context_ptr_val, &otel_span.sc);
+	if (rc != 0) {
+		bpf_printk("failed to write span context: %ld", rc);
+	}
+
+	void *span_ptr_val = get_argument(ctx, 4);
+    bpf_map_update_elem(&active_spans_by_span_ptr, &span_ptr_val, &otel_span, 0);
```

Comment:
I think we should not put un-sampled spans on this map - this will allow us to handle more traffic if sampling is used.

Reply:
Won't that mean that context will not be propagated for un-sampled spans? That would mean things like the parent based sampler wouldn't work correctly, right?

Reply:
If we call `start_tracking_span` it should make sure the context is propagated. That means we need to call `stop_tracking_span` when the span in ended. However, we use the `active_spans_by_span_ptr` to fetch the span in the end probe so not updating the map means we won't be able to fetch the span in the return probe, and won't clean the span context maps properly.

I'd prefer to save maps space for un-sampled spans but it seems more tricky than I initially thought.

Reply:
@RonFed do you have thoughts on how to solve this? I'm not sure how to track span context without also tracking the span.

Should we just open an issue to look into this later?

Reply:
Yeah, worth opening an issue, it is relevant to all the probes.

Reply:
Tracking here: https://github.com/open-telemetry/opentelemetry-go-instrumentation/issues/1117

Reply:
I'm going to resolve this comment as this looks like a systemic issue we want to address and it doesn't seem like something that should block this initial probe for the SDK.

LMK if that is in error.